relayer/chains/cosmos/cosmos_chain_processor.go

```diff
@@ -0,0 +1,554 @@
+						})
+					}
+				}
+			}
+
+			ccp.channelOpenStateLock.Lock()
+			for channelKey, messages := range foundMessages {
+				// do not relay on closed channels
+				if !ccp.channelOpenState[channelKey] {
+					continue
+				}
+				for _, pp := range ccp.pathProcessors {
+					pp.HandleNewMessages(chainID, channelKey, messages)
+				}
+			}
+			ccp.channelOpenStateLock.Unlock()
+			latestQueriedBlock = i
+		}
+
+		if firstTimeInSync {
+			for _, pp := range ccp.pathProcessors {
+				go pp.ScheduleNextProcess(true)
+			}
+		}
+
+		doneWithThisCycle()
+	}
+}
+
+func (ccp *CosmosChainProcessor) IsRelevantChannel(channelKey ibc.ChannelKey) bool {
+	for _, pathProcessor := range ccp.pathProcessors {
+		if pathProcessor.IsRelevantChannel(ccp.ChainProvider.ChainId(), channelKey) {
+			return true
+		}
+	}
+	return false
+}
+
+func (ccp *CosmosChainProcessor) IsRelevantClient(clientID string) bool {
+	for _, pathProcessor := range ccp.pathProcessors {
+		if pathProcessor.IsRelevantClient(clientID) {
+			return true
+		}
+	}
+	return false
+}
+
+func (ccp *CosmosChainProcessor) BootstrapChannels() error {
+	chainID := ccp.ChainProvider.ChainId()
+	uniqueConnectionIDs := []string{}
+	for _, pp := range ccp.pathProcessors {
+		connectionID, err := pp.GetRelevantConnectionID(chainID)
+		if err != nil {
+			return fmt.Errorf("{%s} error getting connection ID: %w\n", chainID, err)
+		}
+		found := false
+		for _, cID := range uniqueConnectionIDs {
+			if connectionID == cID {
+				found = true
+				break
+			}
+		}
+		if !found {
+			uniqueConnectionIDs = append(uniqueConnectionIDs, connectionID)
+		}
+	}
+
+	var eg errgroup.Group
+	for _, connectionID := range uniqueConnectionIDs {
+		connectionID := connectionID
+		eg.Go(func() error {
+			return ccp.UpdateChannelState(connectionID)
+		})
+	}
+	return eg.Wait()
+}
+
+func (ccp *CosmosChainProcessor) UpdateChannelState(connectionID string) error {
+	channels, err := ccp.ChainProvider.QueryConnectionChannels(ccp.ctx, ccp.latestHeight(), connectionID)
+	if err != nil {
+		return err
+	}
+	ccp.channelOpenStateLock.Lock()
+	defer ccp.channelOpenStateLock.Unlock()
+	for _, channel := range channels {
+		ccp.channelOpenState[ibc.ChannelKey{
+			ChannelID:             channel.ChannelId,
+			PortID:                channel.PortId,
+			CounterpartyChannelID: channel.Counterparty.ChannelId,
+			CounterpartyPortID:    channel.Counterparty.PortId,
+		}] = channel.State == chantypes.OPEN
+	}
+	return nil
+}
+
+func (ccp *CosmosChainProcessor) processTransaction(tx *abci.ResponseDeliverTx, errCh chan<- error) []TransactionMessage {
+	messages := []TransactionMessage{}
+	parsedLogs, err := sdk.ParseABCILogs(tx.Log)
+	if err != nil {
+		return messages
+	}
+	for _, messageLog := range parsedLogs {
+		var packetInfo *ibc.PacketInfo
+		var channelInfo *ibc.ChannelInfo
+		var clientInfo *ibc.ClientInfo
+		var action string
+		for _, event := range messageLog.Events {
+			switch event.Type {
```

Comment:
Might be good to log an ignore event type at debug level in the default case.

Reply:
This is a good idea for debugging